<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>stalkless (guide) - automated phylogeny generation for ecologists</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="Will Pearse">

    <!-- Le HTML5 shim, for IE6-8 support of HTML elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Le styles -->
    <link href="assets/css/bootstrap.css" rel="stylesheet">
    <link href="assets/css/bootstrap-responsive.css" rel="stylesheet">
    <link href="assets/css/docs.css" rel="stylesheet">
    <link href="assets/js/google-code-prettify/prettify.css" rel="stylesheet">

    <!-- Le fav and touch icons -->
    <link rel="shortcut icon" href="assets/ico/favicon.ico">
    <link rel="apple-touch-icon" href="assets/ico/apple-touch-icon.png">
    <link rel="apple-touch-icon" sizes="72x72" href="assets/ico/apple-touch-icon-72x72.png">
    <link rel="apple-touch-icon" sizes="114x114" href="assets/ico/apple-touch-icon-114x114.png">
  </head>

  <body data-spy="scroll" data-target=".subnav" data-offset="50">


	  <!-- Navbar
	    ================================================== -->
	    <div class="navbar navbar-fixed-top">
	      <div class="navbar-inner">
	        <div class="container">
	          <a class="btn btn-navbar" data-toggle="collapse" data-target=".nav-collapse">
	            <span class="icon-bar"></span>
	            <span class="icon-bar"></span>
	            <span class="icon-bar"></span>
	          </a>
	          <a class="brand" href="./index.html">stalkless</a>
	          <div class="nav-collapse">
	            <ul class="nav">
	              <li class="">
	                <a href="./index.html">Home</a>
	              </li>
	              <li class="">
	                <a href="./install.html">Install</a>
	              </li>
	              <li class="active">
	                <a href="./guide.html">Guide</a>
	              </li>
	              <li class="">
	                <a href="./faq.html">FAQ</a>
	              </li>
	              <li class="">
	                <a href="http://www.willpearse.com">Who?</a>
	              </li>
	            </ul>
	          </div>
	        </div>
	      </div>
	    </div>

	    <div class="container">

<!-- Masthead
================================================== -->
<header class="jumbotron subhead" id="overview">
  <h1>Guide</h1>
  <p class="lead">How to use stalkless</p>
  <div class="subnav">
    <ul class="nav nav-pills">
	  <li><a href="#overview">Overview</a></li>
	  <li><a href="#method">Method</a></li>
        </ul>
      </li>
    </ul>
  </div>
</header>


<!-- Overview
================================================== -->
<a name="overview"></a>
<section id="overview">
  <div class="page-header">
    <h1>Overview</h1>
  </div>
  <div class="row">
    <div class="span6">
		<p>Once you've got stalkless installed, and you know how to run it (see the 'install' section), give it the arguments <em>-input FOLDER_NAME</em> to show it a folder containing nothing but images to be processed, and <em>-output FOLDER_NAME</em>  to specify where your output should be saved.</p>
    	<p><a href="https://www.dropbox.com/s/6r30ur4f0p5bbbr/demoInput.zip">Try these example files out</a> - on my Mac, I type: <em>./stalkless.app/Contents/MacOS/stalkless -input /Users/will/demoInput -output /Users/will/demoOutput</em>. NOTE: all paths are absolute; that means they have to start with a '/' on Mac and Linux computers, and something like 'C:\' on a Windows computer.</p>
		<p>If you get confused, run the program with only the argument <em>--help</em> to get... help.</p>
		<p>All output is in terms of PIXELS. You can convert this to cm, mm, etc., by knowing the resolution at which each image was scanned.</p>
	</div>
    <div class="span6">
		<p>There are three more arguments you might want to know about:</p>
		<dl>
			<dt>-noObjects</dt>
			<p>Specify how many objects (leaves) are in each image. Tends to improve image detection.</p>
			<dt>-exclusion</dt>
			<p>A comma-separate list (no spaces) of how many pixels to ignore when loading files. In the example, I use this to cut out the information about what the leaves are. This makes it much easier to detect the leaves!</p>
			<dt>-analyseNow</dt>
			<p>Attempt to run the R script the program generates immediately. This will likely crash the program - unless the image segmentation went perfectly, you'll likely have one or two images that are no really leaves. Check you image files, then run the R script.</p>
		</dl>
    </div>
    <div class="span4">
		
    </div>
</section>


<!-- Overview
================================================== -->
<a name="method"></a>
<section id="method">
  <div class="page-header">
    <h1>Methods</h1>
  </div>
  <div class="row">
    <div class="span6">
		<ol>
			<li>Load images, make them black and white. Trim off exclusion zones.</li>
			<li>Thresholds the image - sets anything more intense (dark) than the mean intensity of the image + one standard deviation black, makes the rest white.</li>
			<li>There are a number of unconnected regions of black in the image. Count them, and count how many pixels each one is.</li>
			<li>Ignoring the background, select all regions larger than the mean region size + two standard deviations. These are the leaves. Alternatively, pick the largest X regions as the leaves.</li>
			<li>Count the number of pixels in each leaf. Trim off all white-space surrounding the leaves, then save them out.</li>
			<li>Write out an R script, which contains code to for the Fourier analysis of the leaves and counts the edge length of each leaf.</li>
			<li>Done.</li>
	</div>
    <div class="span6">
		<p>Essentially, the whole code assumes you'll have bimodal image data - that the leaves are much larger and brighter than anything else in the image. If you have extremely small, or extremely faint, leaves, this may not be the case.</p>
		<p>There are much better ways of detecting leaves, which involves more complex edge detection that just thresholding, but this seems to work! If you want details on the interior of the leaves, this is something I'm working on, so talk to me.</p>
    </div>
    <div class="span4">
		
    </div>
</section>


<!-- Start-up
================================================== -->
<a name="startUp"></a>
<section id="startUp">
  <div class="page-header">
    <h1>Start-Up</h1>
  </div>
  <div class="row">
	<div class="span6">
		Double-click 'phyloGenerator.exe' in Windows, or run './phyloGenerator.app/Contents/MacOS/phyloGenerator' from a Mac. On a Mac, <strong>do not double click the application</strong>.
    	<h4>Stem Name</h4>
			<p>phyloGenerator will automatically output all your files, but you need to give it a 'stem name' so that it knows what to call everything. I'd suggest 'test' for your first attempt!</p>
		<h4>Working Directory</h4>
		<p>phyloGenerator needs to know where to put your files. Give a full, absolute path.
		<h4>Gene Name(s)</h4>
		<p>Enter the abbreviated names of the genes you're interested in, one per line, and then hit enter on a blank line to continue. You can also use presets for plants (<em>rbcL</em> and <em>matK</em>), and for animals (<em>COI</em>).</p>
	  <p>Not all sequences on GenBank are annotated in the same way; for example, a cytochrome oxidase one sequence might be labelled as a 'COI', 'Cox1', or 'cytochrome oxidase one' sequence. You can use 'gene aliasing' to account for this, in this example specifying the gene's name as "COI-Cox1-cytochrome_oxidase_one" - aliases are split using '-', and spaces are replaced with '_'.</p>
		<h4>DNA Input</h4>
			<p>If you've already downloaded or sequenced your data, but not aligned it, tell phyloGenerator where that FASTA formatted file is now. Otherwise, just press enter.</p>
			</div>
	    <div class="span6">
			<h4>DNA Download</h4>
				<p>If you don't have DNA sequences already, you'll need to download some. Tell phyloGenerator where a plain text file containing the binomial names of the species you want is. Each species must be on a separate line, with no formatting of any kind in the file. Otherwise, just press enter.</p>
				<p>If you're downloading DNA, you'll need to provide a valid email address so Entrez can contact you about your downloads - this is if you download too much, and has never happened to me.</p>
				<p>You may have to wait some time before your sequences have all been downloaded, but phyloGenerator will tell you how it's doing. By default, it waits ten seconds after downloading each DNA sequence so that it doesn't irritate the servers at GenBank.</p>
				<p>You may see warnings relating to 'NCBI's DTD' files. This is nothing to worry about.</p>
</section>


<!-- DNA Checking
================================================== -->
<a name="dnaChecking"></a>
<section id="dnaChecking">
  <div class="page-header">
    <h1>DNA Checking</h1>
  </div>
  <div class="row">
    <div class="span4">
		<p>This table shows each DNA sequence's 'ID', name used when searching GenBank (if appropriate), the name it was given on GenBank (usually an ID number), and its length. Sequences that are in the top or bottom 5% of sequence lengths are highlighted in the list. Be sensible: if you've downloaded an entire mitochondrial genome (~16000 base pairs) then this won't happily align with a single gene, and you'll crash most of the programs that phyloGenerator uses. Trim those long sequences! Remember that there will always be a 'longest' sequence - but by trimming your sequences you can be sure they're not too long.</p>
   	<p>There are five modes when 'checking' DNA: <em>delete</em>, <em>reload</em>, <em>trim</em>, <em>replace</em>, and <em>merge</em>. You can enter any of the modes by typing their names. To continue, simply hit enter.</p>
	<dl>
		<dt>delete</dt>
		<dd>Enter a single sequence's 'Seq ID' to delete it.</dd>
		<dd>Enter 'gene' and then a gene name to delete that gene entirely. This makes it easy to check to see what genes are available on GenBank for your species, without committing you to use those sequences. Note that you can also run phyloGenerator with the 'nGene' argument to let phyloGenerator select the n genes that give the best coverage of your species from the list of possible genes you provide.</dd>
		<dd>Enter 'output' to write out the sequences you've downloaded so far; but note they will all be outputted at the end of the run anyway.</dl>
		</dl>
	</div>
    <div class="span4">
	<dl>
		<dt>reload</dt>
		<dd>Enter a single sequence's 'Seq ID' to download another sequence from GenBank with the search term originally used. If you put a gene name after the SeqID, e.g. '0rbcL', you'll just reload one gene. You can also use the '>' and '>' operators to reload unusually large/small sequences.</dd>
		<dd>You also have a number of options 'rnd', 'median', 'max', 'min', 'targetLength' that determine how phyloGenerator picks sequences for you from GenBank. For example, if you have an 'nCheck' of 20 and had selected 'median', phyloGenerator will download the first 20 sequences it finds for your species from GenBank (...fewer if there aren't that many!) and use the sequence whose length is closest to the median of them. This can be useful if you're trying to replace your shorter sequences with longer ones.</dl>
		<dt>trim</dt>
		<dd>Enter a single sequence's 'Seq ID' to trim a sequence down, first using any annotations the people who uploaded it to GenBank gave it, and then secondly to its Open Reading Frame (ORF). phyloGenerator will only do one of these things each time you ask it to trim - so if you're still not happy with a gene, try trimming again. However, I would warn that you're unlikely to get a good result from a sequence that wasn't annotated correctly. If you want to trim all the sequences (often a good idea), enter 'EVERYTHING' and press enter. <b>DO NOT</b> attempt to trim a non-coding gene down to its ORF - it doesn't have one!</a> If you trim, and find that you're left with a sequence of length 0, this means phyloGenerator could find neither an ORF nor any annotations on the sequence, and has deleted the sequence (because it's likely rubbish). Consider <em>reload</em>ing that species' sequence.
			<dd>To change the translation table for the sequence you're working with (it might be from a plant's chloroplast, for example), type <em>type</em> and select the appropriate number for each gene.It's important you do this <b>before</b> trimming, because the translation table determines what a sequence's ORF is.</dd>
			</dl>
	</div>
	<div class="row">
    <div class="span4">
		<dl>
			<dt>replace</dt>
			<dd>If there are no sequences for a species on GenBank, type its SeqID to replace it with a congener. Alternatively, type 'EVERYTHING' to replace all missing species with a congener, or 'THOROUGH' to replace all species without any sequences with a relative it is more related to than any other species according to the GenBank taxonomy. This will work even if a species isn't on GenBank at all, but typically not if there's no congener in the database.</dd>
			<dt>merge</dt>
			<dd>If there are no sequences for a species on GenBank, you can merge it/them with another species that do. phyloGenerator will replace the merged species with a polytomy at the end of the run, splitting them half-way along the species with sequence data's branch. Just enter the SeqIDs of the species you want to merge, separated by commas.</dd>
		</dl>
	</div>
</section>


<!-- Alignment and Checking
================================================== -->
<a name="align"></a>
<section id="align">
  <div class="page-header">
    <h1>Alignment</h1>
  </div>
  <div class="row">
	<div class="span4">
    	<h4>Alignment</h4>
				<p>Simply press enter, and phyloGenerator will align your DNA using MAFFT. However, you have a choice of programs: MUSCLE, MAFFT, Clustal-Omega, or PRANK. To use any of them, simply enter <em>muscle</em>, <em>mafft</em>, <em>clustalo</em>, or <em>prank</em> respectively. To compare the alignments of several programs, enter <em>everything</em>, or just to use the first three type <em>quick</em>. Note that PRANK can take a <b>very</b> long time, and I often use 'quick'.</p>
				<p>When you're ready to continue, just hit enter. You'll pick an alignment for each gene (you can only have one alignment) by entering the alignment number for each gene, or if you've only used one alignment method phyloGenerator will just continue straight through.
				<p>Remember you can return to the sequence download stage, or align sequences again, if you're unhappy with what you have.</p>
			</div>
	    <div class="span4">
			<h4>Alignment Checking</h4>
				<p>It is strongly recommended that you look at your alignment before using it. Summary statistics for your alignment(s) have been outputted, and you may notice that one alignment has more gaps than another, or that they all seem OK. You can write out your alignment(s) by typing <em>output</em>. If you wish to return to the alignment or DNA download states, just type <em>DNA</em> or <em>align</em>. There are several other options available to you: <em>trimAl</em>, <em>raxml=X</em> and <em>metal</em>.</p>
				<p>Among all the sequences in the alignments, 'Med. Gaps' is its median number of gaps, 'SD Gaps' the standard deviation of the mean number of gaps, 'Min-Max Gaps' the smallest and largest number of gaps in one sequence, 'Med. Gap Fraction' is the median of the fraction of the sequences made of gaps, 'M-M Gap Fraction' the smallest and largest fraction of sequences with gaps, and 'Warn?' indicates if your alignment is unusually long, given the maximum length of the sequences going into it. Look at these numbers; in particular, warning flags are something to think about.</p>
				</div>
	    <div class="span4">					
				<dl>
					<dt>trimAl</dt>
					<dd><a href="http://trimal.cgenomics.org/">trimAl</a> finds difficult to align regions, and automatically cuts them out. It's a very neat exploratory tool, but a good general rule of thumb is that if you've thrown away a lot of your alignment, the sequences you were using aren't likely to be much good!</dd>
					<dt>raxml=X</dt>
					<dd>Different alignments give different results when run through phylogeny-building programs. RAxML is very quick, and so you can use it to run a number (replace the 'X' with the number of runs you want) of searches and examine the mean Robinson-Folds distance between trees built using the different alignments. Depending on how many alignments and genes you have, this can take quite some time, so do a small run (maybe 5) to guage how long this will take. You'll be shown a distance matrix, where each element is the mean distance between runs.</dd>
					<dt>metal</dt>
					<dd>Comparing sequence alignments is tricky, but <a href="http://kumiho.smith.man.ac.uk/whelan/software/metal/">metal</a> calculates the 'SSP' summary statistic for you and gives a distance matrix, like with RAxML.</dd>
					<dt>clustal-x2</dt>
					<dd>This will open the Clustal X-2 website in your browser, so you can view your alignments more easily if you've outputted them.</dd>
				</dl>
				</div>
</section>

<!-- Constraint Tree
================================================== -->
<a name="constraint"></a>
<section id="constraint">
  <div class="page-header">
    <h1>Constraint Tree</h1>
  </div>
  <div class="row">
    <div class="span4">
			<p>It is <strong>strongly advised</strong> that you use a constraint tree in your analysis. Generating novel phylogenetic hypotheses without a fair amount of knowledge is a tricky thing, so by constraining your tree to fit in with the general consensus of how your group fits together, you're reducing the likelihood that your results will be untenable.</p>
			<p>If you already have a constraint tree, type <em>newick</em>, then tell phyloGenerator where the phylogeny is.</p>
	</div>
    <div class="span4">
	<h4>Building a Constraint Tree</h4>
		<p>To build a constraint tree with Phylomatic, type <em>phylomatic</em>, then give the path to the reference phylogeny and taxonomy when asked. These need to match up perfectly with what you've entered into phyloGenerator, which can be difficult if you've merged taxa. Phylomatic can be a little fiddly, but phyloGenerator attempts to make some changes to the phylogeny to make things run smoothly. You might like to try using Phylomatic outside phyloGenerator if you hit any major problems.</p>
		<p>You can also use GenBank's taxonomy service to find the taxonomy of the species in your phylogeny. Just type 'taxonomy', wait a while, and at the end of your phyloGenerator run a taxonomy of your species will be outputted. You can use this to manually construct a constraint tree, perhaps putting members of the same family together.</p>
	</div>
    <div class="span4">
	<h4>Checking</h4>
		<p>Constraint trees alter the way BEAST and RAxML search for trees, and alter your result. In an ideal world, the genes you've picked convey the 'truth' about the phylogeny of your species, but you might want to know whether that's true. If so, just enter the number of times you want to run a RAxML search both with and without the constraint tree (hit enter to not do so), and you'll be given the mean Robinson-Foulds distance between searches with and without constraints, as well as the variation within the searches.</p>
	</div>
</section>


<!-- Phylogeny Generation
================================================== -->
<a name="phylogeny"></a>
<section id="phylogeny">
  <div class="page-header">
    <h1>Phylogeny Generation</h1>
  </div>
  <div class="row">
    <div class="span4">
	<h4>Overview</h4>
    	<p>This is it! By simply pressing enter, you will make a phylogeny using RAxML. That means you'll generate a maximum likelihood tree, with branch lengths proportional to mutation rates in your genes (you can 'rate smooth' this tree later if you wish). You can alter RAxML's parameters by typing <em>raxml</em>, or build a Bayesian tree with BEAST by typing <em>beast</em>.</p>
    </div>
  
  <div class="row">
    <div class="span4">
	<h4>RAxML</h4>
    	<p>You need to chain the options you want to use together, for example <em>accelBootstrap-noPartitions</em>.</p>
		<p>If you want a good result, you'll have to 'restart' your search a few times (processing power is cheap; why not 100?). Alternatively, you can use the 'accelBootstrap' method, which will conduct multiple searches and also give bootrapped confidences on each of the nodes of your phylogeny; this is my preferred option and tends to be quicker than restarts, so why not do this 200 times?</p>
	</div>
    <div class="row">
      <div class="span4">
  	<h4>BEAST</h4>
      	<p>The defaults (GTR-GAMMA) are likely to be fine for most uses. You'll be using a relaxed log-normal clock, and you won't be estimating any dates or rates of mutation. At this moment in time, phyloGenerator can't handle multiple searches - this will change in the future.</p>
		<p>phyloGenerator does <b>not</b> check MCMC chains for convergence - you will need to do this yourself. This is a deliberate choice; a method that is never wrong does not exist, and I don't want to give people false hope. If someone finds a method (...) I'll incorporate it, and buy them an evening's worth of drinks. If you don't know what any of this means, I suggest you don't use BEAST!</p>
		<p>If you use a constraint tree, phyloGenerator will set strong priors on the ages of all named nodes. For example, the node '(Quercus_robur:30, Quercus_ilex:30)Quercus:44' would be drawn from a normal distribution with a mean of 74 and an SD of 1.</p>
    </div>
</section>

<a name="smoothing"></a>
<section id="smoothing">
  <div class="page-header">
    <h1>Rate Smoothing</h1>
  </div>
  <div class="row">
    <div class="span4">
	<h4>Overview</h4>
		<p>RAxML creates phylogenies whose branch lengths are proportional to the number of mutations that have taken places along them. However, for most purposes, you will want a phylogeny whose branch lengths are proportional to time. We can do this either using PATHd8, which essentially averages out variation in branch lengths (very quickly), or conduct a BEAST search with the topology constrained to what was found by RAxML. All the caveats above about BEAST searches apply to this method of rate smoothing, although I've never had any convergence issues using BEAST in this way.</p> 
	</div>
	<div class="span4">
	<h4>PATHd8</h4>
    	<p>By default, the root age is set to 1, and PATHd8 is run like this. I have no plans to incorproate node-specific ages by default into phyloGenerator - it is probably easier to just do this by hand. I would argue that in my field, community phylogenetics, setting the root age to 1 is better than working with a phylogeny with branch lengths proportional to mutation rate.</p>
		<p>Note that you have to set an outgroup with this method, which is not the case for BEAST.</p>
    </div>
	<h4>BEAST</h4>
    	<p>All the options avaiable to you when performing a normal BEAST run are now available. It might be argued that deviating from the default model choice (GTR-GAMMA) makes no sense, however, as this is (roughly) the model that RAxML uses. As in a normal BEAST search, a log-normal clock is assumed.</p>
    </div>
</section>

<!-- Output
================================================== -->
<a name="output"></a>
<section id="output">
  <div class="page-header">
  <div class="page-header">
    <h1>Output</h1>
	  <div class="row">
	    <div class="span6">
		<h4>End of the line!</h4>
	<p>You're now done! phyloGenerator has produced intermediate files (sequences downloaded, aligned sequences, a log of what was done) and your final phylogeny/phylogenies with the stem name you specified at the beginning.</p>
	<p>Check your phylogeny. Does it look how you'd expect? Are there unusually long branches? Are there polytomies (lots of branches from the same node)? The FAQ has some hints on what to do if you're not happy.</p>
  </div>
    <div class="span6">
	    <h4>Feedback</h4>
	<p>phyloGenerator is very much a work in progress. If there are features you'd like to see in the program, please list them on the feature tracking list (if they're not already there). phyloGenerator does exactly what I want it to do, but you likely want other things. We're happy to give them to you, but we need to know what you need! Thank you!</p>
	<p>In particular, I'm working on more 'sanity checks' for the output of phyloGenerator, so I'd be grateful for any datasets you've run through the program. For the time being, in exchange for letting me see your results, I'll discuss them with you and give you some help!</p>
    </div>
</section>

 <!-- Footer
  ================================================== -->
  <footer class="footer">
    <p class="pull-right"><a href="#">Back to top</a></p>
	<p>&copy <a href="http://www.willpearse.com">Will Pearse</a> (2013), Post-doc in the Cavender-Bares Lab at the University of Minnesota.</p>
    <p>Website (heavily) based on Twitter's Bootstrap - thank you! Bootstrap was built by <a href="http://twitter.com/mdo" target="_blank">@mdo</a> and <a href="http://twitter.com/fat" target="_blank">@fat</a>, their code licensed under the <a href="http://www.apache.org/licenses/LICENSE-2.0" target="_blank">Apache License v2.0</a>.</p>
  </footer>

</div>


    <!-- Le javascript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script type="text/javascript" src="http://platform.twitter.com/widgets.js"></script>
    <script src="assets/js/jquery.js"></script>
    <script src="assets/js/google-code-prettify/prettify.js"></script>
    <script src="assets/js/bootstrap-transition.js"></script>
    <script src="assets/js/bootstrap-alert.js"></script>
    <script src="assets/js/bootstrap-modal.js"></script>
    <script src="assets/js/bootstrap-dropdown.js"></script>
    <script src="assets/js/bootstrap-scrollspy.js"></script>
    <script src="assets/js/bootstrap-tab.js"></script>
    <script src="assets/js/bootstrap-tooltip.js"></script>
    <script src="assets/js/bootstrap-popover.js"></script>
    <script src="assets/js/bootstrap-button.js"></script>
    <script src="assets/js/bootstrap-collapse.js"></script>
    <script src="assets/js/bootstrap-carousel.js"></script>
    <script src="assets/js/bootstrap-typeahead.js"></script>
    <script src="assets/js/application.js"></script>


  </body>
</html>
